admission-policy/readme.adoc

```diff
@@ -0,0 +1,342 @@
+            secretName: opa-ca
+```
+
+```bash
+kubectl create -f opa-admission-controller-deployment.yaml -n opa
+```
+
+When OPA starts, the sidecar (`kube-mgmt`) will register it as an External
+Admission Controller. To verify that registration succeeded, run `kubectl proxy`
+in another terminal and then query the Kubernetes API for the list of External
+Admission Controllers.
+
+```bash
+curl localhost:8001/apis/admissionregistration.k8s.io/v1alpha1/externaladmissionhookconfigurations
+```
+
+Finally, you can follow the OPA logs to see the webhook requests being issued
+by the Kubernetes API server:
+
+```
+kubectl logs -l app=opa -c opa -n opa
+```
+
+== Step 3:  Enforce Kubernetes Admission Control with OPA
+
+=== 3.1 Load a policy into OPA
+To test admission control, create a policy that restricts exec access on
+privileged pods:
+
+**image_source.rego**:
+
+```ruby
+package system
+
+# Deny requests that include container images not from ECR.
+deny[explanation] {
+    image_name = input.spec.object.Spec.Containers[_].Image
+    image_name_parts = split(image_name, "/")
+    repo_name = image_name_parts[0]
+    not startswith(repo_name, "12345678.dkr.ecr.us-west-2.amazonaws.com")
+    explanation = sprintf("image '%v' not from AWS ECR", [image_name])
+}
+
+
+# main is entry point to policy.
+# Boilerplate required by admission webhook.
+# Policy decision is `status`, which is either
+#   {"allowed": true} or
+#   {"allowed": false, "status": {"reason": <string>}}
+main = {
+    "apiVersion": "admission.k8s.io/v1alpha1",
+    "kind": "AdmissionReview",
+    "status": {"allowed": allowed, "status": {"reason": reason}}
+}
+
+# Boilerplate: construct 'reason' and 'allowed' variables.
+#  Real policy is collection of 'deny' statements.
+#  If not denied, allow.
+reason = msg {
+    msg = concat(", ", deny)
+}
+default allowed = true
+allowed = false { n = count(deny); n > 0 }
+
+```
+
+Store the policy in Kubernetes as a ConfigMap.
+
+```bash
+kubectl create configmap privileged-exec --from-file=image_source.rego -n opa
+```
+
+The OPA sidecar will notice the ConfigMap and automatically load the contained
+policy into OPA.
+
+=== 3.2 Check that the policy is working
+
+To verify that your policy is working, create separate test Pods in the `production` namespace.
+
+**nginx-pod.yaml**:
+
+```yaml
+kind: Pod
+apiVersion: v1
+metadata:
+  name: nginx
+  labels:
+    app: nginx
+spec:
+  containers:
+  - image: nginx
+    name: nginx
+```
+
+**amazon-linux-pod.yaml**:
+
+```yaml
+kind: Pod
+apiVersion: v1
+metadata:
+  name: amazon-linux-pod
+  labels:
+    app: amazon-linux
+spec:
+  containers:
+  - image: 12345678.dkr.ecr.us-west-2.amazonaws.com/amazon-linux
+    name: amazon-linux
+```
+
+Create a 'production' namespace.
+```bash
+kubectl create namespace production
+```
+
+Verify that you can create an amazon-linux pod.
+```bash
+kubectl -n production create -f amazon-linux-pod.yaml
+```
+
+Verify that you CANNOT create an nginx pod.
```

Comment:
The goal was to define a policy that prevents users from running `kubectl exec` on "privileged" containers in the "production" namespace. Is it to not be able to create a pod altogether?

Reply:
Editing mismatch here.  The goal of this tutorial is to write admission policies that stop people from using images not coming from an AWS repository.  I corrected the Goal section at the top of the doc.